Valid data masking (#1593)
LiberTEM/LiberTEM: docs/source/udf/advanced.rst

    node, too. Views for aux data are set correctly on the main node. Previously,
    it was only executed on the worker nodes.

Valid data masking
------------------

> In various intermediate states, result buffers may be only partially filled
> with valid data. This comes from the fact that we split the data set into
> partitions, and then merge results together as they become available.
> Also, when doing live processing, you may have a direct mapping of acquired
> detector frames to result items, where naturally only positions where we have
> received and processed data can be valid in the corresponding result buffer.

Reviewer: ```suggestion
In various intermediate states, result buffers may be only partially filled
with valid data. This comes from the fact that we split the dataset into
partitions and then merge results together only as they become available.
When doing live processing we also have the constraint that not all frames
have been acquired yet, and so naturally only positions where we have
received and processed data can be valid in the corresponding result buffer.
```

to access the data of the buffer as a numpy masked array.

For example:

.. testsetup::

    from libertem.udf.sumsigudf import SumSigUDF

.. testcode::

    # use `run_udf_iter` to get partial results while the computation is running:
    for res in ctx.run_udf_iter(dataset=dataset, udf=SumSigUDF()):

        # get the BufferWrapper for the 'intensity' buffer of the first UDF:
        buf = res.buffers[0]['intensity']

        # boolean mask that contains True for valid elements:
        mask = buf.valid_mask

        # get a numpy masked array that is limited to the valid elements:
        masked_data = buf.masked_data

By default, the valid masks of results depend on the buffer kind:

=================== ==================
Buffer :code:`kind` Default valid mask
=================== ==================
:code:`'single'`    All-valid
------------------- ------------------
:code:`'sig'`       All-valid
------------------- ------------------
:code:`'nav'`       Parts which have been merged using :meth:`~libertem.udf.UDF.merge`
=================== ==================

In your UDF, you can also customize the valid mask that should be returned.

> This is useful if there is no direct mapping between the already processed
> navigation elements and the valid results, or if you are using a :code:`kind='single'`
> buffer which is progressively built up, to override the default of all-valid.

Reviewer: ```suggestion
This is useful if the dataset might contain invalid frames discovered at runtime,
or if intermediate results cannot be fully valid due to missing information (e.g.
calculating gradients across parition boundaries). Anothe example is when using a :code:`kind='single'` buffer which is progressively computed, valid-by-default
might be incorrect if not enough information has been seen (e.g. statistical measures
or a count of values).
```

Customizing the valid data mask is done in :meth:`~libertem.udf.UDF.get_results`
with the :meth:`~libertem.udf.UDF.with_mask` method. For example:

.. testsetup::

>     from libertem.udf import UDF

Reviewer: Import (unused, but in the test code)

```suggestion
    from libertem.udf import UDF
    from libertem.common.math import count_nonzero    
```
